API Endpoint and functionality for Centrality calculations

koebi opened the pull request:
### Pull Request Checklist
<!--- Please make sure you have completed the following items BEFORE submitting a pull request (put an x in each box when you have checked you have done them): -->
- [X] 1. I have [**rebased**](https://github.com/GIScience/openrouteservice/blob/master/CONTRIBUTE.md#pull-request-guidelines) the latest version of the master branch into my feature branch and all conflicts have been resolved.
- [X] 2. I have added information about the change/addition to functionality to the CHANGELOG.md file under the [Unreleased] heading.
- [ ] 3. I have documented my code using JDocs tags.
- [X] 4. I have removed unnecessary commented out code, imports and System.out.println statements.
- [X] 5. I have written JUnit tests for any new methods/classes and ensured that they pass.
- [ ] 6. I have created API tests for any new functionality exposed to the API.
- [X] 7. If changes/additions are made to the app.config file, I have added these to the app.config wiki page on github along with a short description of what it is for, and documented this in the Pull Request (below).
- [X] 8. I have built graphs with my code of the Heidelberg.osm.gz file and run the api-tests with all test passing
- [X] 9. I have referenced the Issue Number in the Pull Request (if the changes were from an issue).
- [X] 10. For new features or changes involving building of graphs, I have tested on a larger dataset (at least Germany) and the graphs build without problems (i.e. no out-of-memory errors).
- [x] 11. For new features or changes involving the graphbuilding process (i.e. changing encoders, updating the importer etc.), I have generated longer distance routes for the affected profiles with different options (avoid features, max weight etc.) and compared these with the routes of the same parameters and start/end points generated from the current live ORS. If there are differences then the reasoning for these **MUST** be documented in the pull request.
- [X] 12. I have written in the Pull Request information about the changes made including their intended usage and why the change was needed.

### Information about the changes
- Key functionality added:
Implement API endpoint `centrality` and functionality to calculate betweenness centrality for nodes in a given bounding box.

koebi committed:
Rebase onto master

koebi committed:
Add current state of centrality algorithm development

koebi committed:
current state to push to git

koebi committed:
current state to push to git

koebi committed:
Epic file file store

koebi committed:
Epic file file store

koebi committed:
git is not a file store

koebi committed:
git is not a file store

koebi committed:
rename algorithm since we are implementing Brandes' Algorithm, not FloydWarshall

koebi committed:
rename algorithm since we are implementing Brandes' Algorithm, not FloydWarshall

koebi committed:
finish renaming

koebi committed:
finish renaming

koebi committed:
add basic test, change result type, fix errors in implementation

koebi committed:
add basic test, change result type, fix errors in implementation

koebi committed:
refactor coordinate calculation from node ids to be done outside of algorithm

koebi committed:
refactor coordinate calculation from node ids to be done outside of algorithm

koebi committed:
add twoComponentGraph Test, comment out print debugging statements

koebi committed:
add twoComponentGraph Test, comment out print debugging statements

koebi committed:
add current status

koebi committed:
add current status

koebi committed:
return node Ids for each location

koebi committed:
return node Ids for each location

koebi committed:
Add API Parameter to exclude nodes from centrality request.

koebi committed:
Add API Parameter to exclude nodes from centrality request.

koebi committed:
Remove unnecessary ToDos and debugging statements

koebi committed:
Remove unnecessary ToDos and debugging statements

koebi committed:
Delete old version of main algorithm loop.

Since we're dealing with double, we'll run into rounding errors and need to refactor the main section of the algorithm to include a tolerance for an equality check.

koebi committed:
Delete old version of main algorithm loop.

Since we're dealing with double, we'll run into rounding errors and need to refactor the main section of the algorithm to include a tolerance for an equality check.

koebi committed:
Remove networkx graph export function used for testing.

koebi committed:
Remove networkx graph export function used for testing.

koebi committed:
Merge remote-tracking branch 'origin/centrality' into centrality

# Conflicts:
#	openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfileManager.java

koebi committed:
Add change to changelog

koebi committed:
remove method header of removed method "writeNetworkxGraph"

koebi committed:
remove unused GeoJSON-Response class and import

sfendrich commented on openrouteservice/src/main/java/org/heigit/ors/api/responses/centrality/CentralityResponse.java:
```diff
@@ -0,0 +1,30 @@
+    public CentralityResponse() {};
+
+    // In RouteResponse, this method was used to get metadata from RouteRequest.
+    // The unused CentralityRequest is a remnant of this.
```
If it is not necessary, then please remove the parameter

sfendrich commented on openrouteservice/src/main/java/org/heigit/ors/centrality/algorithms/brandes/BrandesCentralityAlgorithm.java:
```diff
@@ -0,0 +1,148 @@
+            HashMap<Integer, Integer> sigma = new HashMap<>();
+
+            // single source shortest path
+            //S, P, sigma = SingleSourceDijkstra(graph, nodesInBBox, s);
```
This is just a question, nothing to be done: could one of the existing routing algorithms be used here?

koebi replied:
Not without any modification - as far as I know, no routing algorithm keeps track of predecessors on shortest paths or even considers alternative shortest paths of same length.

sfendrich commented on openrouteservice/src/main/java/org/heigit/ors/routing/RoutingProfile.java:
```diff
@@ -660,6 +665,55 @@ public MatrixResult computeMatrix(MatrixRequest req) throws Exception {
+
+        HashMap<Integer, Double> betweenness = alg.compute(nodesInBBox);
+
+        //System.out.printf("Working on %d nodes, result has %d nodes\n, nodesInBBox.size(), betweenness.size());
```
Please remove commented out code. If you need such an output for debugging, consider using an appropriate logger.

sfendrich reviewed (changes requested):
Please also remove unused imports.

koebi committed:
remove unused imports

koebi committed:
remove unused constructor

koebi committed:
remove unused constructor parameter

koebi committed:
Add parameter tests for centrality endpoint

koebi committed:
correctly parse coordinates and throw Exception on error

koebi committed:
add tests for excluding nodes and invalid bBoxes

koebi committed:
rewrite excludeNodes param Test to be graph-agnostic

Currently, we assume some node IDs to correspond to the Theodor-Heuss-Brücke in Heidelberg. Node IDs are not fixed or derived deterministically from the OSM, so updates or re-building of graphs may break this. This commit ensures that the test won't fail in this case.
